### DISCUSSION TITLE
Can't include spdlog as submodule

### DISCUSSION THREAD

**Discussion by DeveloperChaseLewis** [gabime/spdlog#2327] [Q&A]
A bit new to CMake but the instructions in the ReadMe.md don't work because find_package won't work when using a package as a submodule (cmake is so awesome ...). Every other library I am using can just include the header files and the built lib from

`${CMAKE_BINARY_DIR}/submodules/package/src`

Spdlog is the only exception where I don't even get the headers included for some reason I can't quite explain.

Anyone know how to get spdlog to work as a submodule?
```

# CMakeList.txt : CMake project for Valkyrie, include source and define
# project specific logic here.
#
cmake_minimum_required (VERSION 3.8)

# Add source to this project's executable.
find_package(Vulkan REQUIRED)

file(GLOB_RECURSE sources CONFIGURE_DEPENDS "*.cpp")
file(GLOB_RECURSE headers CONFIGURE_DEPENDS "*.h")

add_library(Valkyrie ${headers} ${sources})
target_compile_features(Valkyrie PUBLIC cxx_std_20)

target_include_directories(Valkyrie
	PUBLIC "../submodules/glfw/include"
	PUBLIC "../submodules/glm"
	PUBLIC "../submodules/entt/single_include"
	PUBLIC "../submodules/spdlog/include"
	PUBLIC ${Vulkan_INCLUDE_DIRS}
)

target_link_libraries(Valkyrie
"${CMAKE_BINARY_DIR}/submodules/glfw/src/glfw3.lib"
 ${Vulkan_LIBRARIES}
)
```

**Comment by tt4g** [comment]
Usually, `add_subdirectory()` is used to import another CMake project (e.g. `add_subdirectory(submodules/spdlog)`).
Note that only subdirectories can be included.

See: https://cmake.org/cmake/help/latest/command/add_subdirectory.html

After `add_subdirectory()` imports the spdlog project, an `spdlog::spdlog_header_only` (header-only target) CMake target is defined.

Example:

```diff
+add_subdirectory(submodules/spdlog)
+
 target_include_directories(Valkyrie
         PUBLIC "../submodules/glfw/include"
         PUBLIC "../submodules/glm"
         PUBLIC "../submodules/entt/single_include"
-        PUBLIC "../submodules/spdlog/include"
         PUBLIC ${Vulkan_INCLUDE_DIRS}
 )

 target_link_libraries(Valkyrie
 "${CMAKE_BINARY_DIR}/submodules/glfw/src/glfw3.lib"
  ${Vulkan_LIBRARIES}
+PRIVATE spdlog::spdlog_header_only
 )
```

**Comment by yscontrol** [comment]
Just make sure there is no confusion in terminology, as add_subdirectory
does not IMPORT but rather builds with the project from source. Contrary
add_library(spdlogger SHARED IMPORTED ) does import spdlog into projected
from a prebuilt package without building it from source
